File: mid/Greedy Algorithm/2311. Longest Binary Subsequence Less Than or Equal to K.py
# https://leetcode.com/problems/longest-binary-subsequence-less-than-or-equal-to-k/solutions/2168527/python-greedy-fast-easy-to-undestand-with-explanations/
# Time O(n)
# Space O(1)

class Solution:
    def longestSubsequence(self, s: str, k: int) -> int:
        ones = [i for i, char in enumerate(s[::-1]) if char == '1']
        ans = len(s) - len(ones)
        
        for i in ones:
            if k-2**i < 0:
                break
            k -= 2**i
            ans += 1
        
        return ans

File: mid/Greedy Algorithm/2311. Longest Binary Subsequence Less Than or Equal to K.cpp
// https://www.youtube.com/watch?v=b4oyNbYLSk4
// Time O(n)
// Space O(1)

class Solution {
public:
    int longestSubsequence(string s, int k) {
        int n = s.size(), res = 0, sum = 0;

        for(int i=n-1; i>=0; i--){
            int num = s[i]-'0';
            if(num == 0)
                res++;
            else{
                int j = n-i-1;
                if(sum + pow(2, j) <= k){
                    sum += pow(2, j);
                    res++;
                }
            }
        }
        return res;
    }
};